### PR TITLE
AI-664 subforms

### PR DESCRIPTION
**Pull request by yuriyz**

### CODE REVIEW

File: core/model/src/main/java/org/activityinfo/model/type/period/PeriodValue.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    private int week;
+    private int day;
+    private int hour;
+    private int minute;
+    private int second;
+
+    public PeriodValue() {
+    }
+
+    public PeriodValue(int year, int month, int week, int day, int hour, int minute, int second) {
```
Comment (akbertram) [inline core/model/src/main/java/org/activityinfo/model/type/period/PeriodValue.java]: I think there's a modeling problem here. Values like 'ISO-week of year' or 'month of year' are not "periods of time" as generally understood, they describe an interval between two LocalDates. 

Take a look at Joda-Time's documentation, they have a good presentation of concepts related to chronology, in particular:
http://www.joda.org/joda-time/key.html
http://www.joda.org/joda-time/key_period.html

Take a look at the time types I added on the 3.x branch:
https://github.com/bedatadriven/activityinfo/tree/remaster/core/model/src/main/java/org/activityinfo/model/type/time

The key is that the various types: MonthValue, YearValue all implement an interface TemporalValue which provides asInterval() method which allows us handle the various interval types generally.

File: core/model/src/main/java/org/activityinfo/model/type/subform/ClassType.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+import java.util.Map;
+
+/**
+ * Represents type of form classes : locationType classes or activity classes.
+ * Mainly used to allow user define subform kind.
+ *
+ * @author yuriyz on 03/12/2015.
+ */
+public enum ClassType {
```
Comment (akbertram) [inline core/model/src/main/java/org/activityinfo/model/type/subform/ClassType.java]: Can you explain this a bit? Why is it necessary to distinguish between "types" of FormClasses?

Comment (yuriyz) [inline core/model/src/main/java/org/activityinfo/model/type/subform/ClassType.java]: You can have period subforms like Daily, Weekly. Or subform for class : Partner, Project, Location Type. ClassType represents those subform class types. Is the name confusing?

File: core/model/src/main/java/org/activityinfo/model/type/subform/SubFormType.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+        @Override
+        public FieldValue deserialize(Record record) {
+            return Quantity.fromRecord(record);
+        }
+    }
+
+    public static final TypeClass TYPE_CLASS = new TypeClass();
+
+    private ReferenceType classReference;
```
Comment (akbertram) [inline core/model/src/main/java/org/activityinfo/model/type/subform/SubFormType.java]: Why is this property of type ReferenceType? A subform field is defined by exactly one form class, right? And there will always be a one-to-many relationship between the parent form and the subform? So we could replace this with just 

```
ResourceId subFormId;
```

right?

Comment (yuriyz) [inline core/model/src/main/java/org/activityinfo/model/type/subform/SubFormType.java]: Right. To be honest I don't remember why I've made it ReferenceType. Also I don't see any usages of reference type except getting single id. I will replace it with ResourceId.

File: core/model/src/main/java/org/activityinfo/model/type/subform/ClassType.java (hunk 2 above)
Comment (akbertram) [inline core/model/src/main/java/org/activityinfo/model/type/subform/ClassType.java]: But why handle Partner, Project, Location FormClasses specially? Why wouldn't we just address the general case of "one subform for each resource in a given FormClass" ?

And taking this a step further, whether they're periodic forms or bound to a collection, what they all have in common is a "key" field. For monthly forms it would be a key of type MonthType, if you wanted one entry for each location, the key field would be of type ReferenceType. This would make it easy to expand this to a text field for user-supplied IDs, etc. 

Then basically the only additions we would need to make to the model would be:

```.java
FormClass {
  /* add: */
  Optional<ResourceId> parentFormId;   /* added */
  Optional<FormField> keyField;             /* added */
}
SubFormType : FieldType {
  ResourceId subFormId;
}
```

For SubForm Resources, we could use two reserved words to store the key and the parent id: for example, `__key` and `__parent`

Do you think this would work? 

It would give us a pretty uniform path to dealing with one of the most challenging aspects of the subforms: maintaining the key constraint. We will need to deal with the case of two users submitting a "sub resource" for the same parent and the same key. This gives us a common way of handling all of the cases.

Comment (yuriyz) [inline core/model/src/main/java/org/activityinfo/model/type/subform/ClassType.java]: It is actually already implemented in this way, maybe a bit different. `ClassType` is just convenient way to bind legacy model and new model. It is super-type, ClassType.partner means fetch all partners for given db. Here are a bit more details:

Reference constants : https://github.com/bedatadriven/activityinfo/blob/ai-664/core/model/src/main/java/org/activityinfo/model/type/subform/SubformConstants.java

Subform is described by `FormClass`. Now I'm storing in subform `FormClass` with two special form fields:

1. Reference on value type `FormClass`.

```
 ReferenceType typeClass = (ReferenceType) subForm.getField(SubformConstants.TYPE_FIELD_ID).getType();
 ResourceId typeClassId = typeClass.getRange().iterator().next();
```

Under value type `FormClass` I mean form class for which we will fetch/store instances. Like Partner, Location Type, Period or user-defined form class. Theoretically it can be
any `FormClass` which we support. 

Here is translation from legacy to new model in QueryExecutor (internally check is made by pre-defined ids `hardcoded` in ClassType.).
```
 private Promise<List<FormInstance>> instancesByClassType(ParentCriteria.Parent parent) {
        ResourceId parentId = parent.getParentId();

        if (parentId.equals(ClassType.LOCATION_TYPE.getResourceId())) {
            return locationTypes(parent);
        } else if (parentId.equals(ClassType.PROJECT.getResourceId())) {
            return projects(parent);
        } else if (parentId.equals(ClassType.PARTNER.getResourceId())) {
            return partner(parent);
        } else {
            throw new UnsupportedOperationException("ClassType is not supported, classType: " + parentId);
        }
    }
```

2. Metadata formfield. For example we want to have different tab numbers for specific subform. E.g. Location Type may have long names and we may want to keep 4 tabs, while 
for period it may be 10. 

```
QuantityType tabsCountType = (QuantityType) subForm.getField(SubformConstants.TAB_COUNT_FIELD_ID).getType();
```

In this way we can store in subform `FormClass` any number of metadata we need (for now it is only tab count and value type `FormClass`).

To summarize, current model is generic and we may support any value type `FormClass`.

Comment (yuriyz) [inline core/model/src/main/java/org/activityinfo/model/type/subform/ClassType.java]: Please check also this code snippet, it clarifies how instances for tabs are actually fetched:

https://github.com/bedatadriven/activityinfo/blob/ai-664/server/src/main/java/org/activityinfo/ui/client/component/form/subform/SubFormTabsManipulator.java#L88-88

File: core/model/src/main/java/org/activityinfo/model/type/period/PeriodValue.java (hunk 1 above)
Comment (yuriyz) [inline core/model/src/main/java/org/activityinfo/model/type/period/PeriodValue.java]: @akbertram I'm not sure I really got the problem. Right now we have multi-field representation of period by `PeriodValue` (do not have chronology, it's just for static representation of period). Based on it we convert it to interval (with chronology), so Subform manipulator is using interval as `FormInstance` (two dates). See this:
https://github.com/bedatadriven/activityinfo/blob/ai-664/server/src/main/java/org/activityinfo/ui/client/component/form/subform/InstanceGenerator.java#L94-94

Comment (yuriyz) [inline core/model/src/main/java/org/activityinfo/model/type/period/PeriodValue.java]: @akbertram in this code I've sticked to `java.util.Date` because there is nice `com.google.gwt.user.datepicker.client.CalendarUtil` which helps to manipulate the dates. Do you think we should use `LocalDate` instead? Then we should *invent* own CalendarUtil.
